engines/adl/adl.cpp

```diff
@@ -0,0 +1,1038 @@
+
+namespace Adl {
+
+AdlEngine::~AdlEngine() {
+	delete _display;
+}
+
+AdlEngine::AdlEngine(OSystem *syst, const AdlGameDescription *gd) :
+		Engine(syst),
+		_display(nullptr),
+		_gameDescription(gd),
+		_isRestarting(false),
+		_isRestoring(false),
+		_saveVerb(0),
+		_saveNoun(0),
+		_restoreVerb(0),
+		_restoreNoun(0),
+		_canSaveNow(false),
+		_canRestoreNow(false) {
+}
+
+Common::String AdlEngine::readString(Common::ReadStream &stream, byte until) const {
+	Common::String str;
+
+	while (1) {
+		byte b = stream.readByte();
+
+		if (stream.eos() || stream.err())
+			error("Error reading string");
+
+		if (b == until)
+			break;
+
+		str += b;
+	};
+
+	return str;
+}
+
+Common::String AdlEngine::readStringAt(Common::SeekableReadStream &stream, uint offset, byte until) const {
+	stream.seek(offset);
+	return readString(stream, until);
+}
+
+void AdlEngine::printMessage(uint idx, bool wait) const {
+	Common::String msg = _messages[idx - 1];
+	wordWrap(msg);
+	_display->printString(msg);
+
+	if (wait)
+		delay(14 * 166018 / 1000);
+}
+
+void AdlEngine::delay(uint32 ms) const {
+	Common::EventManager *ev = g_system->getEventManager();
+
+	uint32 start = g_system->getMillis();
+
+	while (!g_engine->shouldQuit() && g_system->getMillis() - start < ms) {
+		Common::Event event;
+		if (ev->pollEvent(event)) {
+			if (event.type == Common::EVENT_KEYDOWN && (event.kbd.flags & Common::KBD_CTRL)) {
+				switch(event.kbd.keycode) {
+				case Common::KEYCODE_q:
+					g_engine->quitGame();
+					break;
+				default:
+					break;
+				}
+			}
+		}
+		g_system->delayMillis(16);
+	}
+}
+
+Common::String AdlEngine::inputString(byte prompt) const {
+	Common::String s;
+
+	if (prompt > 0)
+		_display->printString(Common::String(prompt));
+
+	while (1) {
+		byte b = inputKey();
+
+		if (g_engine->shouldQuit() || _isRestoring)
+			return 0;
+
+		if (b == 0)
+			continue;
+
+		if (b == ('\r' | 0x80)) {
+			s += b;
+			_display->printString(Common::String(b));
+			return s;
+		}
+
+		if (b < 0xa0) {
+			switch (b) {
+			case Common::KEYCODE_BACKSPACE | 0x80:
+				if (!s.empty()) {
+					_display->moveCursorBackward();
+					_display->setCharAtCursor(APPLECHAR(' '));
+					s.deleteLastChar();
+				}
+				break;
+			};
+		} else {
+			if (s.size() < 255) {
+				s += b;
+				_display->printString(Common::String(b));
+			}
+		}
+	}
+}
+
+byte AdlEngine::inputKey() const {
+	Common::EventManager *ev = g_system->getEventManager();
+
+	byte key = 0;
+
+	_display->showCursor(true);
+
+	while (!g_engine->shouldQuit() && !_isRestoring && key == 0) {
+		Common::Event event;
+		if (ev->pollEvent(event)) {
+			if (event.type != Common::EVENT_KEYDOWN)
+				continue;
+
+			if (event.kbd.flags & Common::KBD_CTRL) {
+				if (event.kbd.keycode == Common::KEYCODE_q)
+					g_engine->quitGame();
+				continue;
+			}
+
+			switch (event.kbd.keycode) {
+			case Common::KEYCODE_BACKSPACE:
+			case Common::KEYCODE_RETURN:
+				key = convertKey(event.kbd.keycode);
```

Comment:
I general I *strongly* suggest never ever to mix key codes and ascii values. `convertKey` looks like it works on ascii values not key codes. There is only some overlap. Simply converting the backspace and return keys directly here would be best.